stoppable_workers.go

```diff
@@ -70,13 +104,16 @@ func (sw *StoppableWorkers) Add(worker func(context.Context)) {
 
 // Stop idempotently shuts down all the goroutines we started up.
 func (sw *StoppableWorkers) Stop() {
+	// Call `cancelFunc` with the write lock that competes with "readers" that can add workers. This
+	// guarantees `Add` worker calls that start a goroutine have incremented the `workers` WaitGroup
+	// prior to `Stop` calling `Wait`.
 	sw.mu.Lock()
-	defer sw.mu.Unlock()
```

Comment:
@penguinland I had mentioned before that a worker could not add another worker to the group as it would result in a deadlock. That was because I wrote the `Stop` code slightly incorrectly. Dan fixed it here by making sure `Wait` is not called while holding the mutex.